### DISCUSSION TITLE
关于导出YUV文件的一些问题

### DISCUSSION THREAD

**Discussion by Ansarua** [ryanxingql/mfqev2.0#17] [Q&A]
您好，关于您对自己工作的开源我表示非常感谢。
       我的问题是，当我用utils/file_io.py文件中的import_yuv工具将yuv数据集导入到dataset中，发现仅能返回一帧的Y通道信息。通过only_y这个参数可以修改成分别返回Y,U,V通道的信息，但重点是，分别返回3个通道的信息。我需要的是能不能直接返回channel=3的一个完整的ndarray。
       代码链接如下：https://github.com/ryanxingql/stdf-pytorch/blob/master/dataset/mfqev2.py 190行，这个是import_yuv工具的应用环境
       https://github.com/ryanxingql/stdf-pytorch/blob/master/utils/file_io.py   8行，这个是import_yuv工具的存在位置。
       但我不知道应该如何修改。还希望您能给予一些建议，非常感谢。

**Comment by ryanxingql** [comment]
视频源是 YUV420P 格式，意味着 U、V 分辨率要比 Y 小。此时不可能组成一个 ndarray。
除非你将 U、V 升采样至与 Y 等大，然后才可以组成一个数组。

**Comment by Ansarua** [comment]
视频源是YUV420P格式，那如果我直接用444的方式对它进行输出，生成的YUV能组成一个ndarray吗？

**Comment by ryanxingql** [comment]
> 视频源是YUV420P格式，那如果我直接用444的方式对它进行输出，生成的YUV能组成一个ndarray吗？

如果是444，意味着分辨率相同，可以组成一个ndarray。但前提是你要得到444。你需要升采样UV或降采样Y。

**Comment by Ansarua** [comment]
![1392e4c7ff5d48583efa487739609cb](https://github.com/ryanxingql/mfqev2.0/assets/145305809/35f45ee9-7f6c-441e-8a6f-4eba57ddbfc4)
好的，那您知道出现这种保存结果是什么原因吗？值的范围没调整好？还是什么？

**Comment by ryanxingql** [comment]
色彩通道可能错了。例如 RGB 存成了 BGR。

**Comment by Ansarua** [comment]
![image](https://github.com/ryanxingql/mfqev2.0/assets/145305809/fa0e0620-616d-4e8f-899c-4f602a2a4241)

您好，我尝试着修改过了，可是颜色还是不对，请问还有别的原因会导致这个问题吗？

**Comment by ryanxingql** [comment]
可以看一下完整代码吗？从YUV到RGB的

**Comment by Ansarua** [comment]
这部分并没有代码。我是将444格式的YUV分量直接cat成3通道的ndarray作为网络输入的。后面我察觉到可能是没转成RGB的问题。
![image](https://github.com/ryanxingql/mfqev2.0/assets/145305809/ef3182d7-487e-413f-a6d3-f914dea5ebe4)
请问网上的这个公式是否可用？或者您有YUV转RGB的代码可以分享一下吗？非常感谢

**Comment by ryanxingql** [comment]
可以试试。效果都差不多。建议参考：https://en.wikipedia.org/wiki/YUV

**Comment by Ansarua** [comment]
不太行，改成RGB直接看不到图了。
![image](https://github.com/ryanxingql/mfqev2.0/assets/145305809/23fc5900-4e9d-4abd-9c65-ac85f97e8dc0)

**Comment by Ansarua** [comment]
![image](https://github.com/ryanxingql/mfqev2.0/assets/145305809/45f08fd1-7f92-4df1-9401-595205365ea6)
代码是这样写的。
